### ISSUE TITLE
Add predictive entropy search as acquisition function

### ISSUE THREAD

**Issue by glouppe** [scikit-optimize/scikit-optimize#115]
As suggested by @jmetzen in #19, it would be nice to add entropy search as a new acquisition function, as it often outperforms LCB or EI.

See also https://papers.nips.cc/paper/5324-predictive-entropy-search-for-efficient-global-optimization-of-black-box-functions.pdf

**Comment by MechCoder** [comment]
I'm having a look at the original paper. Hopefully will have something in the next few days.

**Comment by glouppe** [comment]
:+1: 

As far as I remember there is a variant known as "predictive entropy search" which is equivalent but easier to compute.

**Comment by glouppe** [comment]
See https://arxiv.org/abs/1406.2541

**Comment by jmetzen** [comment]
Yes, predictive entropy search would also be interesting since it allows to account for the GP's hyperparameters while entropy search assumes fixed hyperparameters. However, entropy search should be easier to implement (at least the sampling based approach). And you get minimum regret search (https://arxiv.org/abs/1602.01064, shameless plug as I am the author) essentially for free since both are very similar from an implementation point of view. So I would propose to start with implementing entropy search.

**Comment by glouppe** [comment]
Thanks Jan for the helpful feedback! @Manoj: forget my comment then :)

@jmetzen: can we take inspiration from your implementation above?

**Comment by jmetzen** [comment]
Sure, feel free to take inspiration/reuse code whenever it is useful

**Comment by MechCoder** [comment]
I have a question, the paper describes in section 2.2 that the candidate points are sampled from a proposal measure u (for which they say the "probability of improvement' and "expected improvement") are good choices. They also say, for sampling this proposal measure "u", they use the "shrinking rank slice sampling". Is this the same as implemented in your code?

**Comment by jmetzen** [comment]
No, it is not the same. In my code, a kind of (discretized) Thompson sampling is used. I don't think it makes a big difference but I haven't compared it empirically.

**Comment by MechCoder** [comment]
@jmetzen I have been going through http://jmlr.csail.mit.edu/papers/volume13/hennig12a/hennig12a.pdf but I don't understand few things. Here is my understanding as of now.

1. A few representer points are chosen according to some measure (usually sampling from an acquisition function)
2. The `p_min(x_i)` at each of these representer point can be computed using the equation in section 2.3
More simply, for each representer point, we can simply sample from `p(f(x))` N times and `p_min(x_i)` is given by the number of times `f(x_i)` is maximum divided by N.
3. For each new query point, we can compute the change in `mu(x)` and `Sigma(x)` by the formulae given under section 2.4

What I do not understand from here is how to use this information of the change in the mean and Sigma of the new query point to compute the change in `p_min(x_i)` for each of these representer points (and hence the change in entropy)

**Comment by MechCoder** [comment]
Does https://github.com/jmetzen/bayesian_optimization/blob/master/bayesian_optimization/acquisition_functions.py#L319 implement section 2.6?

**Comment by glouppe** [comment]
Do you already have a branch for that Manoj? I am curious to follow more closely :)

**Comment by jmetzen** [comment]
Hi Manoj.

> Does https://github.com/jmetzen/bayesian_optimization/blob/master/bayesian_optimization/acquisition_functions.py#L319 implement section 2.6? (But it says that the derivative of p with respect to the mean and covariance requires expectation propagation)

I only implemented the sampling-based approach and not expectation-propagation (EP). EP would give you derivatives which could help in finding the maximizer of the acquisition function but you can also use gradient-free optimizers like DIRECT or CMA-ES and then you can do without EP.

> What I do not understand from here is how to use this information of the change in the mean and Sigma of the new query point to compute the change in p_min(x_i) for each of these representer points (and hence the change in entropy)   

In principle, you perform sampling again for each assumed change in mean and sigma, get a p_min and use this for computing the change in entropy (information gain). This is of course costly. One useful thing is that the change in the GP's covariance does not depend on the observation sampled from ``p(f(x))`` (at least for a stationary GP). Thus, you can compute generic samples for an assumed observation at ``x``, which use the old mean but the updated covariance. The loop in line 319 does account for the changes in the means for different samples from ``p(f(x))``. 

I should also say that in my implementation, I do not actually sample from the 1d Gaussian ``p(f(x))`` but use percent points (percentiles from the percent point function (inverse of cdf) on an equidistant grid).

HTH, otherwise feel free to ask.

**Comment by glouppe** [comment]
Shall we wait for this to release 0.3 or either release the current master? (this is the last issue marked for the 0.3 milestone)

Given the other improvements, I would vote for preparing the 0.3 release as soon as possible, without waiting for this one.

**Comment by betatim** [comment]
:+1: for a release party.

**Comment by MechCoder** [comment]
So, I have a branch here (https://github.com/scikit-optimize/scikit-optimize/compare/master...MechCoder:entropy_search?expand=1). I'll make a PR and soon as I do a quick verification.

**Comment by betatim** [comment]
Moved this (and #304) to the v0.4 milestone.
